src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/RazorSemanticTokensInfoService.cs

```diff
@@ -58,57 +58,169 @@ internal class RazorSemanticTokensInfoService : IRazorSemanticTokensInfoService
         var tokens = new SemanticTokens { Data = data };
 
+        if (csharpSemanticRanges.HasValue)
+        {
+            csharpSemanticRanges.Value.Dispose();
+        }
+
         return tokens;
     }
 
-    private static List<SemanticRange>? CombineSemanticRanges(List<SemanticRange>? ranges1, List<SemanticRange>? ranges2)
+    private static PooledObject<List<SemanticRange>>? CombineSemanticRanges(List<SemanticRange>? razorRanges, List<SemanticRange>? csharpRanges)
     {
-        if (ranges1 is null || ranges2 is null)
+        if (razorRanges is null || csharpRanges is null)
         {
             // If we have an incomplete view of the situation we should return null so we avoid flashing.
             return null;
         }
 
-        var newList = new List<SemanticRange>(ranges1.Count + ranges2.Count);
-        newList.AddRange(ranges1);
-        newList.AddRange(ranges2);
-
         // Because SemanticToken data is generated relative to the previous token it must be in order.
-        // We have a guarantee of order within any given language server, but the interweaving of them can be quite complex.
-        // Rather than attempting to reason about transition zones we can simply order our ranges since we know there can be no overlapping range.
-        newList.Sort();
+        // We have a guarantee of order within any given language server, but when we translate from the ranges
+        // in the C# document to ranges into the Razor document we lose that guarantee.
+        // Having converted them to SemanticRange objects, we can simply do the final round of a merge sort.
+        var pooledList = ListPool<SemanticRange>.GetPooledObject(out var newList);
+        newList.Capacity = razorRanges.Count + csharpRanges.Count;
+
+        var indexRazor = 0;
+        var indexCsharp = 0;
+        while (indexRazor < razorRanges.Count && indexCsharp < csharpRanges.Count)
+        {
+            var currentRazorRange = razorRanges[indexRazor];
+            var currentCsharpRange = csharpRanges[indexCsharp];
+
+            var comparison = currentRazorRange.CompareTo(currentCsharpRange);
+
+            if (comparison == 0)
+            {
+                // csharp and razor ranges have the same span; skip the C# item
+                ++indexCsharp;
+            }
+            else if (comparison < 0)
+            {
+                newList.Add(currentRazorRange);
+                ++indexRazor;
+            }
+            else
+            {
+                newList.Add(currentCsharpRange);
+                ++indexCsharp;
+            }
+        }
+
+        while (indexRazor < razorRanges.Count)
+        {
+            newList.Add(razorRanges[indexRazor]);
+            ++indexRazor;
+        }
 
-        return newList;
+        while (indexCsharp < csharpRanges.Count)
+        {
+            newList.Add(csharpRanges[indexCsharp]);
+            ++indexCsharp;
+        }
+
+#if DEBUG
```

Comment:
I don't think we should keep this validation, at least not here. If it provides value, then it should be moved to a unit test, which can construct a set of known razor and csharp ranges, call `CombineSemanticRanges`, and validate the result.

It totally makes sense why its valuable during development of this change, but I think if it needs to be permanent, it should be a test.

Reply:
I'm actually concerned that there exists some Razor syntax that would produce out-of-order ranges, and having the Debug code in place lets someone running this in debug to see the issue; otherwise, the ranges are out of order and our simple merge would produce colorization that wouldn't match the text in a significant way.  Will remove later in this release.

Reply:
That concern is fair, but would that not be better placed in the method that just deals with Razor C# ranges? A one-line `Debug.Assert(razorRanges.SequanceEqual(razorRanges.OrderBy()))` would solve that issue, could live in the code forever, and we'd still be able to remove a lot of `#if DEBUG` dead code.

This chunk is _also_ validating that we don't produce overlapping or duplicate tokens, and that seems like its better suited to a unit test.